File: android/app/src/main/java/com/agentcontrolcenter/app/core/config/AgentConfigValidator.kt
package com.agentcontrolcenter.app.core.config

import com.agentcontrolcenter.app.agent.model.AgentConfig
import com.agentcontrolcenter.app.agent.model.AgentType
import com.agentcontrolcenter.app.core.security.UrlValidator

// MARK: - AgentConfigValidator
// 对应 iOS Security/AgentConfigValidator.swift
// 复用 [UrlValidator] 校验 serverUrl，并补齐字段级校验（必填 / 范围 / 长度上限）。
// 在 AgentsViewModel.saveAgent() 入口主动调用，避免无效配置进入 Room。

/**
 * [AgentConfig] 字段校验器。
 *
 * 校验项：
 * - name 非空、长度 ≤ 64
 * - serverUrl 非空、合法 scheme、防 SSRF（复用 [UrlValidator]）
 * - apiKey 非空（LocalModel 类型豁免）
 * - model 非空
 * - temperature ∈ [0.0, 2.0]
 * - maxTokens ∈ [256, 32768]
 * - systemPrompt 长度 ≤ 8000（OpenAI 风格 system 上限，留余量）
 */
object AgentConfigValidator {

    private const val NAME_MAX = 64
    private const val SYSTEM_PROMPT_MAX = 8000
    private const val TEMPERATURE_MIN = 0.0f
    private const val TEMPERATURE_MAX = 2.0f
    private const val MAX_TOKENS_MIN = 256
    private const val MAX_TOKENS_MAX = 32768

    /**
     * 校验 [AgentConfig]。
     *
     * @return [ConfigValidationResult]，errors 为空表示通过
     */
    fun validate(config: AgentConfig): ConfigValidationResult {
        val errors = mutableListOf<ConfigValidationError>()

        // ── name ──
        if (config.name.isBlank()) {
            errors += ConfigValidationError("name", "名称不能为空")
        } else if (config.name.length > NAME_MAX) {
            errors += ConfigValidationError("name", "名称不能超过 $NAME_MAX 字符")
        }

        // ── serverUrl ──
        // LocalModel 走本地进程，serverUrl 不必填；其他类型必须有 URL。
        val isLocal = config.type == AgentType.LocalModel
        if (!isLocal) {
            if (config.serverUrl.isBlank()) {
                errors += ConfigValidationError("serverUrl", "服务器地址不能为空")
            } else {
                val url = UrlValidator.validate(config.serverUrl, allowLocalhost = true)
                if (url == null) {
                    errors += ConfigValidationError(
                        "serverUrl",
                        "地址不合法或不安全（仅支持 http/https/ws/wss，禁止 SSRF 元数据 IP）"
                    )
                }
            }
        }

        // ── apiKey ──
        // LocalModel / OpenCode 等本地部署可豁免；远程协议必须有 apiKey。
        if (!isLocal && config.apiKey.isBlank()) {
            errors += ConfigValidationError("apiKey", "API Key 不能为空")
        }

        // ── model ──
        // LocalModel 也需要 model 名称（Ollama 模型名 / GGUF 文件名等）
        if (config.model.isBlank()) {
            errors += ConfigValidationError("model", "模型名称不能为空")
        }

        // ── temperature ──
        if (config.temperature < TEMPERATURE_MIN || config.temperature > TEMPERATURE_MAX) {
            errors += ConfigValidationError(
                "temperature",
                "temperature 必须在 $TEMPERATURE_MIN ~ $TEMPERATURE_MAX 之间（当前 ${config.temperature}）"
            )
        }

        // ── maxTokens ──
        // OpenAI 上限是 16384/32768 不等，统一用 32768 作上限；下限 256 避免 1-token 截断
        if (config.maxTokens < MAX_TOKENS_MIN || config.maxTokens > MAX_TOKENS_MAX) {
            errors += ConfigValidationError(
                "maxTokens",
                "maxTokens 必须在 $MAX_TOKENS_MIN ~ $MAX_TOKENS_MAX 之间（当前 ${config.maxTokens}）"
            )
        }

        // ── systemPrompt ──
        if (config.systemPrompt.length > SYSTEM_PROMPT_MAX) {
            errors += ConfigValidationError(
                "systemPrompt",
                "系统提示词不能超过 $SYSTEM_PROMPT_MAX 字符（当前 ${config.systemPrompt.length}）"
            )
        }

        return ConfigValidationResult(errors)
    }
}

File: ios/AgentControlCenter/Security/AgentConfigValidator.swift
import Foundation

// MARK: - AgentConfig 校验器
// 对应 Android com.agentcontrolcenter.app.core.config.AgentConfigValidator
//
// 纯静态 `enum` 命名空间，无实例状态，天然线程安全。调用方在保存 AgentConfig
// 前调用 `validate(_:)`，校验失败时根据返回的 `ConfigValidationResult` 回填
// 表单错误并阻止落库。

/// `AgentConfig` 校验器。
///
/// 校验项（与 Android `AgentConfigValidator` 一致）：
/// - `name` 非空，长度 ≤ 64
/// - `serverUrl` 非空 + `URLValidator` 校验（`LocalModel` 类型豁免）
/// - `apiKey` 非空（`LocalModel` 豁免）
/// - `model` 非空
/// - `temperature` ∈ [0.0, 2.0]
/// - `maxTokens` ∈ [256, 32768]
/// - `systemPrompt` 长度 ≤ 8000
enum AgentConfigValidator {

    /// 字段名常量，便于 UI 按 `errorFor(_:)` 检索对应字段的错误消息。
    enum Field {
        static let name = "name"
        static let serverUrl = "serverUrl"
        static let apiKey = "apiKey"
        static let model = "model"
        static let temperature = "temperature"
        static let maxTokens = "maxTokens"
        static let systemPrompt = "systemPrompt"
    }

    /// 校验 AgentConfig。
    /// - Parameter config: 待校验的 Agent 配置
    /// - Returns: 校验结果；`isValid` 为 true 时可安全保存
    static func validate(_ config: AgentConfig) -> ConfigValidationResult {
        var errors: [ConfigValidationError] = []

        // name 非空，长度 ≤ 64
        let trimmedName = config.name.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedName.isEmpty {
            errors.append(ConfigValidationError(field: Field.name, message: "名称不能为空"))
        } else if config.name.count > 64 {
            errors.append(ConfigValidationError(field: Field.name, message: "名称长度不能超过 64 个字符"))
        }

        // serverUrl：非空 + URLValidator（LocalModel 豁免，本地模型走 LocalModelManager）
        if config.type != .localModel {
            let trimmedUrl = config.serverUrl.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmedUrl.isEmpty {
                errors.append(ConfigValidationError(field: Field.serverUrl, message: "服务器地址不能为空"))
            } else if URLValidator.validate(trimmedUrl) == nil {
                errors.append(ConfigValidationError(field: Field.serverUrl, message: "服务器地址不合法或存在安全风险"))
            }
        }

        // apiKey：非空（LocalModel / ComfyUI 豁免）
        // 与 Android AgentConfigValidator 对齐：ComfyUI 本地部署通常无认证，可豁免 apiKey。
        // OpenWebUI 仍需 apiKey（与 OpenAI 一致，远程 HTTP+SSE 协议）。
        let apiKeyOptional = config.type == .localModel || config.type == .comfyUI
        if !apiKeyOptional {
            if config.apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                errors.append(ConfigValidationError(field: Field.apiKey, message: "API Key 不能为空"))
            }
        }

        // model 非空
        if config.model.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            errors.append(ConfigValidationError(field: Field.model, message: "模型不能为空"))
        }

        // temperature ∈ [0.0, 2.0]（AgentConfig.temperature 为 Float）
        if config.temperature < 0.0 || config.temperature > 2.0 {
            errors.append(ConfigValidationError(field: Field.temperature, message: "温度必须在 0.0 ~ 2.0 之间"))
        }

        // maxTokens ∈ [256, 32768]
        if config.maxTokens < 256 || config.maxTokens > 32768 {
            errors.append(ConfigValidationError(field: Field.maxTokens, message: "最大 Tokens 必须在 256 ~ 32768 之间"))
        }

        // systemPrompt 长度 ≤ 8000
        if config.systemPrompt.count > 8000 {
            errors.append(ConfigValidationError(field: Field.systemPrompt, message: "System Prompt 长度不能超过 8000 个字符"))
        }

        return .of(errors)
    }
}
